lib/cloud_controller/perm/client.rb

```diff
@@ -83,6 +83,31 @@ def unassign_roles(org_ids: [], space_ids: [], user_id:, issuer:)
         end
       end
 
+      def has_any_permission?(permissions:, user_id:, issuer:)
+        if enabled
+          permissions.any? do |p|
+            has_permission?(permission_name: p[:permission_name], resource_id: p[:resource_id], issuer: issuer, user_id: user_id)
+          end
+        else
+          false
+        end
+      end
+
+      def has_permission?(permission_name:, resource_id:, user_id:, issuer:)
+        if enabled
+          begin
+            client.has_permission?(actor_id: user_id, issuer: issuer, permission_name: permission_name, resource_id: resource_id)
+          rescue GRPC::BadStatus => e
```

Comment:
This might be valid. Will check with isobel to see if we are supposed to be capturing the other errors here too.

Reply:
Errors for all the error codes (including `Ok` apparently?!) [subclass `BadStatus`](https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/errors.rb). You can take Perm down and CAPI will keep chugging along.

For instance, here are the CAPI logs for unassigning a user from an org role when Perm is completely down:

```
{"timestamp":1512770681.6979861,"message":"Started GET \"/v2/info\" for user: , ip: 127.0.0.1 with vcap-request-id: 4a21bf5b-cab0-4602-b727-d49d45743d10 at 2017-12-08 22:04:41 UTC","log_level":"info","source":"cc.api","data":{"request_guid":"4a21bf5b-cab0-4602-b727-d49d45743d10"},"thread_id":46983040348500,"fiber_id":46983039855340,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":12,"method":"call"}
{"timestamp":1512770681.6992152,"message":"Completed 200 vcap-request-id: 4a21bf5b-cab0-4602-b727-d49d45743d10","log_level":"info","source":"cc.api","data":{"request_guid":"4a21bf5b-cab0-4602-b727-d49d45743d10"},"thread_id":46983040348500,"fiber_id":46983039855340,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":24,"method":"call"}
{"timestamp":1512770685.8480186,"message":"Started GET \"/v2/config/feature_flags/unset_roles_by_username\" for user: 2bc27264-96d3-49d2-9b41-5d1bdc2a25d0, ip: 204.15.0.254 with vcap-request-id: 0a403d2a-768f-4525-6911-f64f92028e20::e5de0f71-54ca-4f0d-a55e-2e5da53a56c8 at 2017-12-08 22:04:45 UTC","log_level":"info","source":"cc.api","data":{"request_guid":"0a403d2a-768f-4525-6911-f64f92028e20::e5de0f71-54ca-4f0d-a55e-2e5da53a56c8"},"thread_id":46983045430340,"fiber_id":46983040760200,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":12,"method":"call"}
{"timestamp":1512770685.8511934,"message":"Completed 200 vcap-request-id: 0a403d2a-768f-4525-6911-f64f92028e20::e5de0f71-54ca-4f0d-a55e-2e5da53a56c8","log_level":"info","source":"cc.api","data":{"request_guid":"0a403d2a-768f-4525-6911-f64f92028e20::e5de0f71-54ca-4f0d-a55e-2e5da53a56c8"},"thread_id":46983045430340,"fiber_id":46983040760200,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":24,"method":"call"}
{"timestamp":1512770686.1563034,"message":"Started GET \"/v2/organizations?q=name%3Aorg&inline-relations-depth=1\" for user: 2bc27264-96d3-49d2-9b41-5d1bdc2a25d0, ip: 204.15.0.254 with vcap-request-id: 0445ee0f-4c9b-4784-68f6-7788af530b43::67bf74ea-e68b-440a-a4d9-3e9d96018c69 at 2017-12-08 22:04:46 UTC","log_level":"info","source":"cc.api","data":{"request_guid":"0445ee0f-4c9b-4784-68f6-7788af530b43::67bf74ea-e68b-440a-a4d9-3e9d96018c69"},"thread_id":46983040346500,"fiber_id":46983041463280,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":12,"method":"call"}
{"timestamp":1512770686.171005,"message":"Completed 200 vcap-request-id: 0445ee0f-4c9b-4784-68f6-7788af530b43::67bf74ea-e68b-440a-a4d9-3e9d96018c69","log_level":"info","source":"cc.api","data":{"request_guid":"0445ee0f-4c9b-4784-68f6-7788af530b43::67bf74ea-e68b-440a-a4d9-3e9d96018c69"},"thread_id":46983040346500,"fiber_id":46983041463280,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":24,"method":"call"}
{"timestamp":1512770686.4802492,"message":"Started DELETE \"/v2/organizations/a2554e88-5b0d-496a-b179-0f3a302c923f/managers\" for user: 2bc27264-96d3-49d2-9b41-5d1bdc2a25d0, ip: 204.15.0.254 with vcap-request-id: c81f9d77-721e-4a14-4422-79d71303e649::6d64dc6c-a7ce-48c2-b3d7-34bb46bfe6db at 2017-12-08 22:04:46 UTC","log_level":"info","source":"cc.api","data":{"request_guid":"c81f9d77-721e-4a14-4422-79d71303e649::6d64dc6c-a7ce-48c2-b3d7-34bb46bfe6db"},"thread_id":46983040346640,"fiber_id":46983041642620,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":12,"method":"call"}
{"timestamp":1512770686.6030643,"message":"unassign-role.bad-status","log_level":"error","source":"perm.client","data":{"request_guid":"c81f9d77-721e-4a14-4422-79d71303e649::6d64dc6c-a7ce-48c2-b3d7-34bb46bfe6db","role":"org-manager-a2554e88-5b0d-496a-b179-0f3a302c923f","user_id":"19984632-867e-4a52-8324-faf299c6795f","issuer":"https://uaa.thread-kicker.perm.cf-app.com/oauth/token","status":"GRPC::Unavailable","code":14,"details":"Connect Failed","metadata":{}},"thread_id":46983040346640,"fiber_id":46983041642620,"process_id":6189,"file":"/var/vcap/data/packages/cloud_controller_ng/4a56beb24965b4f25176cff13eb1b0332ef7cf80/cloud_controller_ng/lib/cloud_controller/perm/client.rb","lineno":147,"method":"rescue in unassign_role"}
{"timestamp":1512770686.6095796,"message":"Completed 204 vcap-request-id: c81f9d77-721e-4a14-4422-79d71303e649::6d64dc6c-a7ce-48c2-b3d7-34bb46bfe6db","log_level":"info","source":"cc.api","data":{"request_guid":"c81f9d77-721e-4a14-4422-79d71303e649::6d64dc6c-a7ce-48c2-b3d7-34bb46bfe6db"},"thread_id":46983040346640,"fiber_id":46983041642620,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":24,"method":"call"}
{"timestamp":1512770688.0304806,"message":"Started GET \"/internal/v4/syslog_drain_urls?batch_size=50&next_id=0\" for user: , ip: 127.0.0.1 with vcap-request-id: 8676d4a8-bf47-405f-96bd-40a8164e4118 at 2017-12-08 22:04:48 UTC","log_level":"info","source":"cc.api","data":{"request_guid":"8676d4a8-bf47-405f-96bd-40a8164e4118"},"thread_id":46983040348260,"fiber_id":46983039506580,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":12,"method":"call"}
{"timestamp":1512770688.0344238,"message":"Completed 200 vcap-request-id: 8676d4a8-bf47-405f-96bd-40a8164e4118","log_level":"info","source":"cc.api","data":{"request_guid":"8676d4a8-bf47-405f-96bd-40a8164e4118"},"thread_id":46983040348260,"fiber_id":46983039506580,"process_id":6189,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb","lineno":24,"method":"call"}
```

Reply:
Are there any other errors that the perm client can return? Should `perm-rb` have it's own error class rather than leaking GRPC errors into it's consumers?

Reply:
No to the first, yes to the second but that's out of scope for this PR

Reply:
https://www.pivotaltracker.com/story/show/153580833